File: prevoz/src/main/java/org/prevoz/android/push/PushFragment.java
package org.prevoz.android.push;

import android.os.Bundle;
import android.support.v7.app.AlertDialog;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import org.prevoz.android.PrevozFragment;
import org.prevoz.android.R;
import org.prevoz.android.events.Events;
import org.prevoz.android.model.NotificationSubscription;
import org.prevoz.android.ui.DividerItemDecoration;
import org.prevoz.android.util.LocaleUtil;
import org.prevoz.android.util.ViewUtils;

import butterknife.ButterKnife;
import butterknife.InjectView;
import de.greenrobot.event.EventBus;
import rx.android.schedulers.AndroidSchedulers;

public class PushFragment extends PrevozFragment
{
    public static final String DIALOG_ARG_SUB = "subscription";

    @InjectView(R.id.notifications_list)
    protected RecyclerView notificationList;

    @InjectView(R.id.empty_view)
    protected View emptyView;


    // Used for dialog callback due to unvieldy interface
    private NotificationSubscription clickedSubscription;

    private PushNotificationsAdapter adapter;

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        View views = inflater.inflate(R.layout.fragment_notifications, container, false);
        ButterKnife.inject(this, views);

        pushManager.getSubscriptions()
                    .observeOn(AndroidSchedulers.mainThread())
                    .subscribe(notifications -> {
                        ViewUtils.setupEmptyView(notificationList, emptyView, "Niste prijavljeni na nobena obvestila.");
                        adapter = new PushNotificationsAdapter(getActivity(), notifications, item -> onItemClicked(item));

                        notificationList.setAdapter(adapter);
                        notificationList.addItemDecoration(new DividerItemDecoration(getActivity(), DividerItemDecoration.VERTICAL_LIST));
                        LinearLayoutManager llm = new LinearLayoutManager(getActivity(), LinearLayoutManager.VERTICAL, false);
                        notificationList.setLayoutManager(llm);
                        updateEmptyView();
                    });


        return views;
    }

    protected void onItemClicked(NotificationSubscription item)
    {
        clickedSubscription = item;
        new AlertDialog.Builder(getActivity(), R.style.Prevoz_Theme_Dialog)
                        .setTitle(String.format("%s - %s", item.getFrom().toString(), item.getTo().toString()))
                        .setMessage(String.format("Ali se res želite odjaviti od obveščanja v %s?", LocaleUtil.getNotificationDayName(getResources(), item.getDate()).toLowerCase()))
                        .setNegativeButton("Prekliči", null)
                        .setPositiveButton("Odjavi", (dialog, which) -> {
                            if (clickedSubscription != null)
                            {
                                pushManager.setSubscriptionStatus(getActivity(), clickedSubscription.getFrom(), clickedSubscription.getTo(), clickedSubscription.getDate(), false);
                            }
                        }).show();
    }

    @Override
    public void onPause()
    {
        super.onPause();
        EventBus.getDefault().unregister(this);
    }

    @Override
    public void onResume()
    {
        super.onResume();
        EventBus.getDefault().register(this);
    }

    private void updateEmptyView() {
        notificationList.setVisibility(adapter.getItemCount() == 0 ? View.INVISIBLE : View.VISIBLE);
        emptyView.setVisibility(adapter.getItemCount() == 0 ? View.VISIBLE : View.INVISIBLE);
    }

    public void onEventMainThread(Events.NotificationSubscriptionStatusChanged e)
    {
        pushManager.getSubscriptions()
                   .observeOn(AndroidSchedulers.mainThread())
                   .subscribe(notifications -> {
                       adapter.setData(notifications);
                       updateEmptyView();
                   });

    }
}

File: prevoz/src/main/kotlin/org/prevoz/android/push/PushFragment.kt
package org.prevoz.android.push

import android.os.Bundle
import android.support.transition.TransitionManager
import android.support.v7.app.AlertDialog
import android.support.v7.widget.LinearLayoutManager
import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import butterknife.BindView
import butterknife.ButterKnife
import com.hannesdorfmann.mosby.mvp.MvpFragment
import org.prevoz.android.PrevozApplication
import org.prevoz.android.R
import org.prevoz.android.model.NotificationSubscription
import org.prevoz.android.model.Route
import org.prevoz.android.util.LocaleUtil
import org.prevoz.android.util.ViewUtils

class PushFragment : MvpFragment<PushFragment, PushPresenter>() {

    @BindView(R.id.notifications_list)
    lateinit var notificationList: RecyclerView

    @BindView(R.id.empty_view)
    lateinit var emptyView: ViewGroup

    override fun createPresenter(): PushPresenter {
        return PushPresenter((activity!!.application as PrevozApplication).component())
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        val views = inflater.inflate(R.layout.fragment_notifications, container, false)
        ButterKnife.bind(this, views as View)
        ViewUtils.setupEmptyView(notificationList, emptyView, "Niste prijavljeni na nobena obvestila.")
        val layoutManager = LinearLayoutManager(context, LinearLayoutManager.VERTICAL, false)
        notificationList.layoutManager = layoutManager
        return views
    }

    fun showNotifications(notifications: List<NotificationSubscription>) {
        TransitionManager.beginDelayedTransition(emptyView)
        if (notifications.isEmpty()) {
            notificationList.visibility = View.INVISIBLE
            emptyView.visibility = View.VISIBLE
        } else {
            notificationList.visibility = View.VISIBLE
            emptyView.visibility = View.INVISIBLE
            notificationList.adapter = PushNotificationsAdapter(context, notifications, {
                subscription -> showNotificationRemoveDialog(subscription)
            })
        }
    }

    private fun showNotificationRemoveDialog(subscription : NotificationSubscription) {
        AlertDialog.Builder(activity!!, R.style.Prevoz_Theme_Dialog)
                .setTitle(String.format("%s - %s", subscription.from.toString(), subscription.to.toString()))
                .setMessage(String.format("Ali se res želite odjaviti od obveščanja v %s?", LocaleUtil.getNotificationDayName(resources, subscription.date).toLowerCase()))
                .setNegativeButton("Prekliči", null)
                .setPositiveButton("Odjavi") { dialog, which ->
                    presenter.unsubscribeFrom(subscription)
                }.show()
    }
}